routers/private/hook_pre_receive.go

```diff
@@ -472,25 +472,41 @@ func (ctx *preReceiveContext) loadPusherAndPermission() bool {
 		return true
 	}
 
-	user, err := user_model.GetUserByID(ctx, ctx.opts.UserID)
-	if err != nil {
-		log.Error("Unable to get User id %d Error: %v", ctx.opts.UserID, err)
-		ctx.JSON(http.StatusInternalServerError, private.Response{
-			Err: fmt.Sprintf("Unable to get User id %d Error: %v", ctx.opts.UserID, err),
-		})
-		return false
-	}
-	ctx.user = user
-
-	userPerm, err := access_model.GetUserRepoPermission(ctx, ctx.Repo.Repository, user)
-	if err != nil {
-		log.Error("Unable to get Repo permission of repo %s/%s of User %s: %v", ctx.Repo.Repository.OwnerName, ctx.Repo.Repository.Name, user.Name, err)
-		ctx.JSON(http.StatusInternalServerError, private.Response{
-			Err: fmt.Sprintf("Unable to get Repo permission of repo %s/%s of User %s: %v", ctx.Repo.Repository.OwnerName, ctx.Repo.Repository.Name, user.Name, err),
-		})
-		return false
+	if ctx.opts.UserID == user_model.ActionsUserID {
+		ctx.user = user_model.NewActionsUser()
+		ctx.userPerm.AccessMode = perm_model.AccessMode(ctx.opts.ActionPerm)
+		if err := ctx.Repo.Repository.LoadUnits(ctx); err != nil {
+			log.Error("Unable to get User id %d Error: %v", ctx.opts.UserID, err)
+			ctx.JSON(http.StatusInternalServerError, private.Response{
+				Err: fmt.Sprintf("Unable to get User id %d Error: %v", ctx.opts.UserID, err),
+			})
+			return false
+		}
+		ctx.userPerm.Units = ctx.Repo.Repository.Units
+		ctx.userPerm.UnitsMode = make(map[unit.Type]perm_model.AccessMode)
+		for _, u := range ctx.Repo.Repository.Units {
+			ctx.userPerm.UnitsMode[u.Type] = ctx.userPerm.AccessMode
+		}
```

Comment:
how about merge these logic into ``GetUserRepoPermission``?

Reply:
I think it's a good idea, but considering `GetUserRepoPermission` is used in many places(50+), I prefer to touch its code in another PR if there is no bug now.

Reply:
`GetUserRepoPermission` has been used in many places but I don't think all these places need this logic.

Reply:
> `GetUserRepoPermission` has been used in many places but I don't think all these places need this logic.

Hmm, at least can add a new function named as `LoadActionsUserPermission()` to do it. otherwise that's a little duplicate. and it will be better to add a option for each repo to control whether allow `ActionsUser` write. example in github:
![image](https://user-images.githubusercontent.com/25342410/215392778-4c913f5e-0ec8-493f-9fda-21b0d895ff3b.png)

Reply:
> and it will be better to add a option for each repo to control whether allow `ActionsUser` write.

You may notice that GitHub has many options for actions permission control in `<owner>/<repo>/settings/actions`:

- Actions permissions
- Fork pull request workflows from outside collaborators
- Workflow permissions
- Fork pull request workflows (for private repo only)
- Access (for private repo only)

So there is lots of work to do, and the plan in my mind is just grant permissions by default now, to make users try it easily. Then use more PRs to provide options to limit permissions. That's why "It shouldn't be used in a public Gitea instance currently".